File: custom_components/family_assistant/frontend/task-settlements.js
/* Explicit consent for ordinary one-off child-task settlement only. */
export const SETTLEMENT_COPY = {
  en: {
    title: "Daily missed-task settlement (optional)",
    hint: "Only one-off child tasks with a deadline. Not personal, recurring, school or maintenance work. Old missed days are skipped without penalties after an outage. Identity, deadline or module changes require a fresh policy review.",
    daily_rollover: "Move an unfinished task to the next day",
    settle_time: "Settle no earlier than (household time)",
    repeat_penalty: "Allow the configured penalty again on later missed days (subject to the global switch and daily cap)",
    same_day_correction: "Reverse only this task's exact penalty when a parent confirms completion on the settlement day",
    review: "Review and reauthorize this policy", corrected: "Correct this task's exact same-day penalty",
    reason: "Parent independently reviewed this task's same-day penalty", needs_review: "Task completed; its exact penalty still needs parent review.",
    receipt: "Last settlement", active: "Active", revoked: "Revoked — review required", disabled: "Off",
    settlement_capacity: "Daily rollover paused: all 366 audit slots are used. History is preserved; no further rollover or penalty. Parent review is required.",
    applied: "Penalty recorded", skipped_outage: "Older day skipped; no penalty", skipped_retroactive: "Pre-policy deadline; no penalty", skipped_disabled: "Automatic penalties or Court off; no penalty", skipped_zero: "No penalty configured", skipped_duplicate: "Already assessed; no duplicate penalty", skipped_cap: "Daily cap reached; no penalty",
  },
  ru: {
    title: "Ежедневный перенос невыполненной задачи (необязательно)",
    hint: "Только разовая задача ребёнка со сроком. Не личные напоминания, серии, школа или обслуживание. После простоя старые дни пропускаются без штрафов. Изменение личности, срока или модулей требует новой проверки политики.",
    daily_rollover: "Переносить невыполненную задачу на следующий день",
    settle_time: "Подводить итог не раньше (время семьи)",
    repeat_penalty: "Разрешить настроенный штраф повторно за последующие пропущенные дни (с общим выключателем и дневным лимитом)",
    same_day_correction: "Отменять только точный штраф этой задачи, если родитель подтвердил выполнение в день начисления",
    review: "Проверить и заново разрешить политику", corrected: "Исправить точный штраф этой задачи за день выполнения",
    reason: "Родитель независимо проверил точный штраф задачи за день выполнения", needs_review: "Задача выполнена; её точный штраф ещё требует проверки родителем.",
    receipt: "Последний перенос", active: "Включено", revoked: "Отозвано — нужна проверка", disabled: "Выключено",
    settlement_capacity: "Ежедневный перенос остановлен: заполнены все 366 записей журнала. История сохранена; новых переносов и штрафов нет. Нужна проверка родителя.",
    applied: "Штраф начислен", skipped_outage: "Старый день пропущен без штрафа", skipped_retroactive: "Срок до включения политики; без штрафа", skipped_disabled: "Автоштрафы или Суд выключены; без штрафа", skipped_zero: "Штраф не настроен", skipped_duplicate: "Уже учтено; без повторного штрафа", skipped_cap: "Дневной лимит достигнут; без штрафа",
  },
  uk: {
    title: "Щоденне перенесення невиконаного завдання (необов’язково)",
    hint: "Лише разове завдання дитини з терміном. Не особисті нагадування, серії, школа чи обслуговування. Після простою старі дні пропускаються без штрафів. Зміна особи, терміну чи модулів потребує нової перевірки політики.",
    daily_rollover: "Переносити невиконане завдання на наступний день",
    settle_time: "Підбивати підсумок не раніше (час родини)",
    repeat_penalty: "Дозволити налаштований штраф повторно за наступні пропущені дні (із загальним вимикачем і денним лімітом)",
    same_day_correction: "Скасовувати лише точний штраф цього завдання, якщо батьки підтвердили виконання в день нарахування",
    review: "Перевірити й повторно дозволити політику", corrected: "Виправити точний штраф цього завдання за день виконання",
    reason: "Батьки незалежно перевірили точний штраф завдання за день виконання", needs_review: "Завдання виконано; його точний штраф ще потребує перевірки батьками.",
    receipt: "Останнє перенесення", active: "Увімкнено", revoked: "Відкликано — потрібна перевірка", disabled: "Вимкнено",
    settlement_capacity: "Щоденне перенесення зупинено: заповнено всі 366 записів журналу. Історію збережено; нових перенесень і штрафів немає. Потрібна перевірка батьків.",
    applied: "Штраф нараховано", skipped_outage: "Старий день пропущено без штрафу", skipped_retroactive: "Термін до ввімкнення політики; без штрафу", skipped_disabled: "Автоштрафи або Суд вимкнено; без штрафу", skipped_zero: "Штраф не налаштовано", skipped_duplicate: "Уже враховано; без повторного штрафу", skipped_cap: "Денний ліміт досягнуто; без штрафу",
  },
};

export const defaultMissedPolicy = () => ({daily_rollover: false, settle_time: "20:00", repeat_penalty: false, same_day_correction: false});
export const settlementCopy = card => SETTLEMENT_COPY[card._config?.language || card._hass?.language?.split("-")[0]] || SETTLEMENT_COPY.en;
export const settlementActorRevision = card => card._data.members.find(m => m.id === card._data.actor)?.revision;
export const settlementSupported = (card, item) => card.parent && !item.delivery_scope && !item.source && !item.series_id && !item.occurrence_id && item.managed_by !== "school" && card._data.members.some(m => m.id === item.assignee && m.active && m.role === "child");

export function settlementControls(card, el, container, initial, onChange) {
  const copy = settlementCopy(card);
  const policy = {...defaultMissedPolicy(), ...initial};
  const fieldset = el("fieldset");
  fieldset.dataset.taskSettlement = "true";
  fieldset.append(el("legend", copy.title), el("p", copy.hint));
  const inputs = {};
  for (const key of Object.keys(policy)) {
    const label = el("label", copy[key]);
    const input = el("input");
    input.name = `missed_${key}`;
    input.type = key === "settle_time" ? "time" : "checkbox";
    if (input.type === "checkbox") input.checked = policy[key];
    else input.value = policy[key];
    label.append(input);
    fieldset.append(label);
    inputs[key] = input;
  }
  const read = () => Object.fromEntries(Object.entries(inputs).map(([key, input]) => [key, input.type === "checkbox" ? input.checked : input.value]));
  const sync = () => {
    if (!inputs.daily_rollover.checked) {
      inputs.repeat_penalty.checked = false;
      inputs.same_day_correction.checked = false;
    }
    for (const key of ["settle_time", "repeat_penalty", "same_day_correction"]) inputs[key].disabled = !inputs.daily_rollover.checked;
    inputs.settle_time.required = inputs.daily_rollover.checked;
  };
  for (const input of Object.values(inputs)) input.addEventListener("change", () => {
    if (fieldset.disabled || card._writing) return;
    sync();
    onChange(read());
  });
  sync();
  container.append(fieldset);
  return {fieldset, read, available(value) { sync(); fieldset.disabled = !value; }, sync};
}

File: custom_components/family_assistant/domain/task_settlements.py
"""Opt-in one-off child-task settlements, committed with their exact score receipt.

No transport I/O, retroactive catch-up loop, or member-name based score lookup.
The original receipt survives correction; only its separate decision changes.
"""

from copy import deepcopy
from datetime import timedelta
from zoneinfo import ZoneInfo

from ..const import PRIVILEGED
from . import penalties, recurrence, task_access
from .context import Context
from .validation import DomainError, fields, revision, text, timestamp

DEFAULTS = {
    "daily_rollover": False,
    "settle_time": "20:00",
    "repeat_penalty": False,
    "same_day_correction": False,
}
OPEN = {"assigned", "accepted", "in_progress", "needs_changes"}
EVENTS = {"task_rollover"}
MAX_SETTLEMENT_RECEIPTS = 366


def supported(state, item):
    member = state["members"].get(item.get("assignee"), {})
    return (
        not task_access.private_task(item)
        and not item.get("source")
        and not item.get("series_id")
        and not item.get("occurrence_id")
        and member.get("active") is True
        and member.get("role") == "child"
        and member.get("revision") == item.get("assignee_revision")
    )


def configure(ctx, item, value):
    ctx.require_parent()
    if not isinstance(value, dict):
        raise DomainError("invalid_field", "missed_policy")
    fields(value, set(DEFAULTS), set(DEFAULTS))
    policy = deepcopy(value)
    for key in ("daily_rollover", "repeat_penalty", "same_day_correction"):
        if type(policy[key]) is not bool:
            raise DomainError("invalid_field", "missed_policy")
    recurrence.clock(policy["settle_time"])
    if not supported(ctx.state, item) or not item.get("due_at"):
        raise DomainError("invalid_field", "missed_policy")
    if policy["daily_rollover"] and len(item.get("missed_receipts", {})) >= MAX_SETTLEMENT_RECEIPTS:
        raise DomainError("invalid_transition", "settlement_capacity")
    if not policy["daily_rollover"] and (policy["repeat_penalty"] or policy["same_day_correction"]):
        raise DomainError("invalid_field", "missed_policy")
    if policy == item.get("missed_policy") and current(ctx.state, item):
        return
    revoke(ctx, item)
    item["missed_policy"] = policy
    item["missed_generation"] = item.get("missed_generation", 0) + 1
    item.setdefault("missed_original_due_at", item["due_at"])
    item["missed_scope"] = {
        "generation": item["missed_generation"],
        "policy": deepcopy(policy),
        "actor": ctx.actor_id,
        "actor_revision": ctx.actor["revision"],
        "member": item["assignee"],
        "member_revision": ctx.member(item["assignee"])["revision"],
        "timezone": ctx.state["settings"].get("timezone", "UTC"),
        "effective_at": ctx.now.isoformat(),
        "state": "active" if policy["daily_rollover"] else "disabled",
    }


def managed(item):
    return item.get("missed_policy", {}).get("daily_rollover") is True


def current(state, item, *, active=True):
    scope = item.get("missed_scope", {})
    actor = state["members"].get(scope.get("actor"), {})
    return (
        "tasks" in state["settings"]["modules"]
        and supported(state, item)
        and managed(item)
        and (not active or scope.get("state") == "active")
        and scope.get("generation") == item.get("missed_generation")
        and scope.get("policy") == item.get("missed_policy")
        and scope.get("member") == item.get("assignee")
        and scope.get("member_revision") == item.get("assignee_revision")
        and scope.get("timezone") == state["settings"].get("timezone", "UTC")
        and actor.get("active") is True
        and actor.get("role") in PRIVILEGED
        and scope.get("actor_revision") == actor.get("revision")
    )


def revoke(ctx, item):
    if isinstance(item.get("missed_scope"), dict):
        # Preserve the truthful terminal capacity reason across scheduler ticks
        # and unrelated revocations. Only explicit reconfiguration replaces it.
        if item["missed_scope"].get("state") != "needs_review":
            item["missed_scope"]["state"] = "revoked"
    for event in ctx.state["outbox"].values():
        if (
            event["key"] in EVENTS
            and event["data"].get("id") == item["id"]
            and event["state"] in {"pending", "awaiting_channel"}
        ):
            event["state"] = "superseded"


def _at(day, deadline, zone):
    result = recurrence.local_clock(day, deadline.strftime("%H:%M"), zone)
    return (
        result + timedelta(seconds=deadline.second, microseconds=deadline.microsecond)
        if result
        else None
    )


def _settle_at(item, deadline, zone):
    cutoff = recurrence.local_clock(deadline.date(), item["missed_policy"]["settle_time"], zone)
    return (
        max(deadline + timedelta(minutes=item["deadline_policy"]["grace_minutes"]), cutoff)
        if cutoff
        else None
    )


def _award(ctx, item, receipt):
    config = ctx.state["settings"]
    if "court" not in config["modules"] or not config.get("automatic_penalties"):
        return "skipped_disabled", None
    points = item["deadline_policy"]["penalty"]
    if points >= 0:
        return "skipped_zero", None
    repeat = item["missed_policy"]["repeat_penalty"]
    # A policy toggle, edited deadline, or reassignment cannot create another
    # score for this task on the same day (including a previously reversed row).
    for event in ctx.state["court"].values():
        if (
            event.get("source") != "task"
            or event.get("reason_data", {}).get("task_id") != item["id"]
        ):
            continue
        day = (
            timestamp(event["created_at"], "created_at")
            .astimezone(ZoneInfo(receipt["timezone"]))
            .date()
            .isoformat()
        )
        if not repeat or day == receipt["settled_local_date"]:
            return "skipped_duplicate", None
    source_id = f"{item['id']}:missed:{receipt['due_local_date']}" if repeat else item["id"]
    record_id = f"task:{source_id}"
    if record_id in ctx.state["court"]:
        return "skipped_duplicate", None
    applied = penalties.award(
        ctx,
        source="task",
        source_id=source_id,
        member=receipt["member"],
        points=points,
        reason_key="task_missed",
        reason_data={
            "task_id": item["id"],
            "due_at": receipt["from_due_at"],
            "settlement_id": receipt["id"],
            "due_local_date": receipt["due_local_date"],
            "member_revision": receipt["member_revision"],
        },
        timezone=receipt["timezone"],
    )
    return ("applied", record_id) if applied else ("skipped_cap", None)


def tick(ctx, item):
    """Called instead of the legacy escalation, only for explicitly managed work."""
    if not current(ctx.state, item) or item["status"] not in OPEN:
        return
    scope = item["missed_scope"]
    zone = scope["timezone"]
    local = ctx.now.astimezone(ZoneInfo(zone))
    deadline = timestamp(item["due_at"], "due_at").astimezone(ZoneInfo(zone))
    scheduled = _settle_at(item, deadline, zone)
    if scheduled is not None and ctx.now < scheduled:
        return
    if scheduled is None and ctx.now < deadline + timedelta(
        minutes=item["deadline_policy"]["grace_minutes"]
    ):
        return  # A nonexistent spring cutoff cannot move a later deadline early.
    if len(item.get("missed_receipts", {})) >= MAX_SETTLEMENT_RECEIPTS:
        revoke(ctx, item)
        scope.update(state="needs_review", reason="settlement_capacity")
        ctx.touch(item)
        return
    start = max(deadline.date() + timedelta(days=1), local.date())
    target = None
    for offset in range(3):
        candidate = _at(start + timedelta(days=offset), deadline, zone)
        candidate_settlement = _settle_at(item, candidate, zone) if candidate else None
        if candidate_settlement and candidate_settlement > ctx.now:
            target = candidate
            break
    if target is None:
        return
    from . import task_events

    task_events.close(ctx, item)
    receipt = {
        "id": f"{item['id']}:missed:{scope['generation']}:{deadline.date().isoformat()}",
        "task_id": item["id"],
        "generation": scope["generation"],
        "policy": deepcopy(scope["policy"]),
        "actor": scope["actor"],
        "actor_revision": scope["actor_revision"],
        "member": scope["member"],
        "member_revision": scope["member_revision"],
        "timezone": zone,
        "due_local_date": deadline.date().isoformat(),
        "settled_local_date": local.date().isoformat(),
        "settled_at": ctx.now.isoformat(),
        "from_due_at": item["due_at"],
        "due_at": target.isoformat(),
        "same_day_correction": item["missed_policy"]["same_day_correction"],
        "reviewers": task_events.reviewers(ctx.state),
    }
    if scheduled is None or scheduled.date() < local.date():
        outcome, ledger_id = "skipped_outage", None
    elif deadline <= timestamp(scope["effective_at"], "effective_at"):
        outcome, ledger_id = "skipped_retroactive", None
    else:
        outcome, ledger_id = _award(ctx, item, receipt)
    receipt.update(outcome=outcome, ledger_id=ledger_id)
    if ledger_id:
        receipt["ledger_revision"] = ctx.state["court"][ledger_id]["revision"]
        receipt["points"] = ctx.state["court"][ledger_id]["points"]
    else:
        receipt["points"] = 0
    item["due_at"] = receipt["due_at"]
    item["missed_receipt"] = receipt
    item.pop("missed_correction", None)
    notifier = Context(ctx.state, ctx.actor, ctx.now, f"task-settlement:{receipt['id']}")
    receipt["event_id"] = notifier.notify(
        "parents",
        "task_rollover",
        {
            "id": item["id"],
            "settlement_id": receipt["id"],
            "member": receipt["member"],
            "member_revision": receipt["member_revision"],
            "due_at": receipt["due_at"],
            "points": receipt["points"],
            "outcome": outcome,
        },
    )
    item.setdefault("missed_receipts", {})[receipt["id"]] = deepcopy(receipt)
    ctx.touch(item)


def _candidate(state, item):
    receipt = item.get("missed_receipt", {})
    if (
        item.get("status") != "completed"
        or not current(state, item, active=False)
        or receipt.get("generation") != item.get("missed_generation")
        or receipt.get("member") != item.get("assignee")
        or receipt.get("member_revision") != item.get("assignee_revision")
        or not receipt.get("same_day_correction")
        or not receipt.get("ledger_id")
        or timestamp(item["closed_at"], "closed_at")
        .astimezone(ZoneInfo(receipt["timezone"]))
        .date()
        .isoformat()
        != receipt["settled_local_date"]
    ):
        return None
    return receipt


def correct(ctx, item, *, payload=None):
    """One exact automatic event; ordinary appeal review rules still apply."""
    ctx.require_parent()
    receipt = _candidate(ctx.state, item)
    if receipt is None:
        if payload is not None:
            raise DomainError("invalid_transition")
        return
    if payload is not None:
        if revision(payload["actor_revision"]) != ctx.actor["revision"]:
            raise DomainError("conflict")
        if payload["settlement_id"] != receipt["id"]:
            raise DomainError("conflict")
        expected = revision(payload["court_revision"])
        reason = text(payload["reason"], "reason", 500)
    else:
        expected = receipt["ledger_revision"]
        reason = "Parent confirmed task completion on the settlement day"
    record = ctx.state["court"].get(receipt["ledger_id"], {})
    data = record.get("reason_data", {})
    if (
        record.get("source") != "task"
        or record.get("actor") != "system"
        or record.get("member") != receipt["member"]
        or record.get("points") != receipt["points"]
        or data.get("task_id") != item["id"]
        or data.get("settlement_id") != receipt["id"]
        or data.get("due_at") != receipt["from_due_at"]
        or data.get("due_local_date") != receipt["due_local_date"]
        or data.get("member_revision") != receipt["member_revision"]
    ):
        raise DomainError("conflict")
    decision = {
        "settlement_id": receipt["id"],
        "ledger_id": record["id"],
        "actor": ctx.actor_id,
        "actor_revision": ctx.actor["revision"],
        "closed_at": item["closed_at"],
        "at": ctx.now.isoformat(),
    }
    if record.get("status") == "reversed":
        decision["state"] = "already_reversed"
    elif "court" not in ctx.state["settings"]["modules"] or expected != record.get("revision"):
        if payload is not None:
            raise DomainError("conflict")
        decision["state"] = "needs_review"
    else:
        from . import court

        try:
            court.handle(
                ctx, "reverse", {"id": record["id"], "revision": expected, "reason": reason}
            )
        except DomainError as error:
            if payload is not None or error.code != "forbidden":
                raise
            decision["state"] = "needs_review"
        else:
            record["reversal"]["source_reference"] = {
                "source": "task",
                "task_id": item["id"],
                "settlement_id": receipt["id"],
            }
            decision["state"] = "reversed"
    item["missed_correction"] = decision


def correction_payload(state, item, actor):
    receipt = _candidate(state, item)
    if not receipt:
        raise DomainError("invalid_transition")
    return {
        "id": item["id"],
        "revision": item["revision"],
        "actor_revision": actor["revision"],
        "settlement_id": receipt["id"],
        "court_revision": state["court"][receipt["ledger_id"]]["revision"],
    }


def event_current(state, event):
    item = state["tasks"].get(event.get("data", {}).get("id"), {})
    receipt = item.get("missed_receipt", {})
    from .task_events import reviewers

    return (
        event.get("key") == "task_rollover"
        and event.get("recipient") == "parents"
        and current(state, item)
        and item.get("status") in OPEN
        and receipt.get("generation") == item.get("missed_generation")
        and receipt.get("event_id") == event.get("id")
        and receipt.get("reviewers") == reviewers(state)
        and event["data"].get("settlement_id") == receipt.get("id")
        and event["data"].get("member") == receipt.get("member")
        and event["data"].get("member_revision") == receipt.get("member_revision")
        and event["data"].get("due_at") == receipt.get("due_at") == item.get("due_at")
        and event["data"].get("points") == receipt.get("points")
        and event["data"].get("outcome") == receipt.get("outcome")
    )


def project(result, *, parent):
    scope = result.pop("missed_scope", None)
    result.pop("missed_generation", None)
    result.pop("missed_receipts", None)
    receipt = result.pop("missed_receipt", None)
    correction = result.pop("missed_correction", None)
    if scope:
        result["missed_policy_status"] = scope["state"]
        if parent and scope.get("reason"):
            result["missed_policy_reason"] = scope["reason"]
    if parent and receipt:
        result["last_settlement"] = {
            key: receipt[key]
            for key in (
                "id",
                "due_local_date",
                "settled_at",
                "from_due_at",
                "due_at",
                "points",
                "outcome",
                "ledger_id",
            )
        }
        result["correction_status"] = correction["state"] if correction else None
